src/Microsoft.ML.TorchSharp/Utils/ThrowHelper.cs

```diff
@@ -0,0 +1,464 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+// This file defines an internal static class used to throw exceptions in BCL code.
+// The main purpose is to reduce code size.
+//
+// The old way to throw an exception generates quite a lot IL code and assembly code.
+// Following is an example:
+//     C# source
+//          throw new ArgumentNullException(nameof(key), SR.ArgumentNull_Key);
+//     IL code:
+//          IL_0003:  ldstr      "key"
+//          IL_0008:  ldstr      "ArgumentNull_Key"
+//          IL_000d:  call       string System.Environment::GetResourceString(string)
+//          IL_0012:  newobj     instance void System.ArgumentNullException::.ctor(string,string)
+//          IL_0017:  throw
+//    which is 21bytes in IL.
+//
+// So we want to get rid of the ldstr and call to Environment.GetResource in IL.
+// In order to do that, I created two enums: ExceptionResource, ExceptionArgument to represent the
+// argument name and resource name in a small integer. The source code will be changed to
+//    ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key, ExceptionResource.ArgumentNull_Key);
+//
+// The IL code will be 7 bytes.
+//    IL_0008:  ldc.i4.4
+//    IL_0009:  ldc.i4.4
+//    IL_000a:  call       void System.ThrowHelper::ThrowArgumentNullException(valuetype System.ExceptionArgument)
+//    IL_000f:  ldarg.0
+//
+// This will also reduce the Jitted code size a lot.
+//
+// It is very important we do this for generic classes because we can easily generate the same code
+// multiple times for different instantiation.
+//
+
+using System.Diagnostics;
+
+namespace System
+{
+#pragma warning disable MSML_GeneralName // This name should be PascalCased
+
+    internal static class ThrowHelper
```

Comment:
Interesting! I'm assuming this is a copy-paste from .NET Core but I'm curious what the net impact of this was? Smaller DLL/Memory footprint? How substantial?

@ericstj  do you know?

Reply:
Yeah, smaller IL.  Looks like smaller JIT'ed size as well.  The above comment summarizes it.  It dates back [quite some time](https://referencesource.microsoft.com/#mscorlib/system/throwhelper.cs,73669ffe9b1ee4f4,references).  You could probably extrapolate net IL savings by taking the above call savings (21 - 7 = 14) and multiplying by usage 1575 in System.Private.Corelib ~ 20 KB. 

Not sure how valuable this is here, if we only call it a couple times it's probably not worth it, just replace those with explicit throw statements.

Reply:
I thought the resolution here was to just delete this?

Reply:
Its been deleted now. Not sure how I missed this comment the first time.